drivers/serial/uart_rtt.c

```diff
@@ -72,49 +86,139 @@ static void uart_rtt_poll_out(struct device *dev, unsigned char c)
 	SEGGER_RTT_Write(ch, &c, 1);
 }
 
+#ifdef CONFIG_UART_ASYNC_API
+
+static int uart_rtt_callback_set(struct device *dev, uart_callback_t callback,
+				 void *user_data)
+{
+	get_dev_data(dev)->callback = callback;
+	get_dev_data(dev)->user_data = user_data;
+	return 0;
+}
+
+static int uart_rtt_tx(struct device *dev, const uint8_t *buf, size_t len,
+		       int32_t timeout)
+{
+	const struct uart_rtt_config *cfg = get_dev_config(dev);
+	struct uart_rtt_data *data = get_dev_data(dev);
+	unsigned int ch = cfg ? cfg->channel : 0;
+
+	ARG_UNUSED(timeout);
+
+	/* Output the buffer */
+	SEGGER_RTT_Write(ch, buf, len);
```

Comment:
`SEGGER_RTT_Write` is taking a lock (mutex in case of zephyr). This can limit places where this driver can be used.
Does Zephyr allow calling uart_tx from atomic context?

Reply:
I suspect that its not defined, and probably won't be until #18970 is resolved.
The nrfx and sam0 drivers are the only ones that currently implement `UART_ASYNC_API`.
They both look like they will work/return `-EBUSY` with interrupts disabled.
I will update to manually take the lock.

Reply:
Updated

Reply:
Not a bad idea but it will cover only cases when uart_tx is called from under irq_lock. It won't cover the case when it is called from an interrupt (mutex cannot be taken from an interrupt).

Maybe it would be enough to leave `SEGGER_RTT_Write` and add a check blocking usage from atomic context?

there is no is_atomic api in Zephyr but check can be done as
```
        int key = irq_lock();                                                                                            
        irq_unlock(key);                                                                                                 
                                                                                                                         
        bool is_atomic = arch_irq_unlocked(key) || k_is_in_isr();
        if (is_atomic) return -ENOTSUP;
```

Reply:
I'd rather not fail when IRQ's are locked if the mutex take would still succeed.
Is adding the following sufficient?
```
if (k_in_in_isr()) {
    return -ENOTSUP;
}
```

Reply:
That's ok. But I strongly suggest adding a check against calling from an interrupt. If anybody does it at least problem will be visible and explicit.

Reply:
Doesn't this setup cover the cases?
If in ISR the `k_in_in_isr` will fail.
If interrupts are locked in a thread the TX will pass if mutex is free, fail if it is in use.
Otherwise it will pass.
Is there some other situation I'm missing?

Reply:
No, it's fine :)